### PR TITLE
SONARJS-882 Rule: 'Symbol' should not be used as a constructor

### PR DESCRIPTION
**Pull request by ivandalbosco**

### CODE REVIEW

File: javascript-checks/src/test/resources/checks/symbolUsedAsConstructor.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
+var sym = Symbol("foo");               // OK
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/symbolUsedAsConstructor.js]: @ivandalbosco It would be nice to have a separate file:
```
class Symbol {} // user-defined "Symbol" in global scope

var s = new Symbol(); // Ok
```

File: javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3834.html [hunk 2]
```diff
@@ -0,0 +1,11 @@
+<p><code>Symbol</code> is an object wrapper and meant to be used like a function. Using it with the <code>new</code> operator will raise a
```
Comment (vilchik-elena) [inline javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S3834.html]: @ivandalbosco Could you update description?

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/Tree.java [hunk 3]
```diff
@@ -992,6 +992,12 @@
     public Class<? extends Tree> getAssociatedInterface() {
       return associatedInterface;
     }
+
+    @Override
+    public boolean contains(TreeKindComposite other) {
+      return this == other;
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/Tree.java]: MAJOR Change this comparison to use the equals method. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1698)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/TreeKinds.java [hunk 4]
```diff
@@ -58,27 +60,41 @@
...
+  }
 
-  private TreeKinds() {
-    // This class has only static methods
+  @Override
+  public boolean contains(TreeKindComposite other) {
+    return equals(other) || subkinds.contains(other);
   }
 
+  // TODO remove
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/TreeKinds.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/TreeKinds.java [hunk 5]
```diff
@@ -58,27 +60,41 @@
     Kind.GENERATOR_DECLARATION,
     Kind.GET_METHOD,
     Kind.SET_METHOD,
-    Kind.ARROW_FUNCTION
-  };
+    Kind.ARROW_FUNCTION);
+
+  private List<Kind> subkinds;
+
+  private TreeKinds(Kind... kinds) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/TreeKinds.java]: MINOR "private" is redundant in this context. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS2333)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/TreeKinds.java [hunk 6]
```diff
@@ -58,27 +60,41 @@
...
+  public boolean contains(TreeKindComposite other) {
+    return equals(other) || subkinds.contains(other);
   }
 
+  // TODO remove
   public static boolean isAssignment(Tree tree) {
     return tree.is(ASSIGNMENT_KINDS);
   }
 
+  // TODO remove
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/TreeKinds.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/TreeKinds.java [hunk 7]
```diff
@@ -58,27 +60,41 @@
...
   public static boolean isIncrementOrDecrement(Tree tree) {
     return tree.is(INC_DEC_KINDS);
   }
 
+  // TODO remove
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/TreeKinds.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/TreeKinds.java [hunk 8]
```diff
@@ -58,27 +60,41 @@
...
   public static boolean isFunction(Tree tree) {
     return tree.is(FUNCTION_KINDS);
   }
 
+  // TODO remove
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/TreeKinds.java]: INFO Complete the task associated to this TODO comment. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1135)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/TreeKindComposite.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.javascript.api.tree;
+
+@FunctionalInterface
+public interface TreeKindComposite {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/TreeKindComposite.java]: MAJOR Drop this interface in favor of "java.util.function.Predicate<TreeKindComposite>". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1711)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 2 major
* ![MINOR][MINOR] 1 minor
* ![INFO][INFO] 4 info

Watch the comments in this conversation to review them.

[INFO]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png 'Severity: INFO'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: javascript-checks/src/main/java/org/sonar/javascript/checks/SymbolUsedAsConstructorCheck.java [hunk 10]
```diff
@@ -55,19 +56,11 @@ private boolean isSymbol(ExpressionTree expression) {
     }
     return false;
   }
-
+  
   private boolean isSymbolShadowed(Tree tree) {
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/SymbolUsedAsConstructorCheck.java]: @ivandalbosco can we undo this change?

File: javascript-checks/src/test/resources/checks/symbolUsedAsConstructor2.js [hunk 11]
```diff
@@ -0,0 +1,5 @@
+var sym = new Symbol(123);   // FN
```
Comment (vilchik-elena) [inline javascript-checks/src/test/resources/checks/symbolUsedAsConstructor2.js]: @ivandalbosco it's not clear from the test why we have this FN

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/Kinds.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.javascript.api.tree;
+
+@FunctionalInterface
+public interface Kinds {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/Kinds.java]: MAJOR Drop this interface in favor of "java.util.function.Predicate<Kinds>". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1711)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/EqualInForLoopTerminationCheck.java [hunk 13]
```diff
@@ -183,15 +181,15 @@ private static boolean isUpdateIncDec(ExpressionTree update) {
...
+    } else if (update.is(KindSet.INC_DEC_KINDS) || update.is(PLUS_ASSIGNMENT, MINUS_ASSIGNMENT)) {
       result = true;
     }
 
     return result;
   }
 
   private static boolean isUpdateOnOneWithAssign(ExpressionTree update) {
-    if (update.is(Tree.Kind.PLUS_ASSIGNMENT, Tree.Kind.MINUS_ASSIGNMENT)) {
+    if (update.is(PLUS_ASSIGNMENT, Tree.Kind.MINUS_ASSIGNMENT)) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/EqualInForLoopTerminationCheck.java]: MINOR Replace this fully qualified name with "MINUS_ASSIGNMENT" [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1942)

File: javascript-checks/src/main/java/org/sonar/javascript/checks/EqualInForLoopTerminationCheck.java [hunk 14]
```diff
@@ -168,7 +166,7 @@ private static Integer getInteger(ExpressionTree expression) {
   }
 
   private static int checkForUpdateByOne(ExpressionTree update) {
-    if (update.is(Tree.Kind.POSTFIX_INCREMENT, Tree.Kind.PREFIX_INCREMENT) || (update.is(Tree.Kind.PLUS_ASSIGNMENT) && isUpdateOnOneWithAssign(update))) {
+    if (update.is(Tree.Kind.POSTFIX_INCREMENT, Tree.Kind.PREFIX_INCREMENT) || (update.is(PLUS_ASSIGNMENT) && isUpdateOnOneWithAssign(update))) {
       return +1;
     }
     if (update.is(Tree.Kind.POSTFIX_DECREMENT, Tree.Kind.PREFIX_DECREMENT) || (update.is(Tree.Kind.MINUS_ASSIGNMENT) && isUpdateOnOneWithAssign(update))) {
```
Comment (SonarTech) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/EqualInForLoopTerminationCheck.java]: MINOR Replace this fully qualified name with "MINUS_ASSIGNMENT" [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1942)

**Comment by SonarTech** [comment]
SonarQube analysis reported 3 issues
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 2 minor

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* ![MAJOR][MAJOR] 1 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
